ragna/core/_rag.py

```diff
@@ -237,42 +278,64 @@ def _parse_documents(self, documents: Iterable[Any]) -> list[Document]:
 
     def _unpack_chat_params(
         self, params: dict[str, Any]
-    ) -> dict[tuple[Type[Component], str], dict[str, Any]]:
-        source_storage_models = self.source_storage._protocol_models()
-        assistant_models = self.assistant._protocol_models()
+    ) -> dict[Callable, dict[str, Any]]:
+        component_models = {
+            getattr(component, name): model
+            for component in [self.source_storage, self.assistant]
+            for (_, name), model in component._protocol_models().items()
+        }
 
         ChatModel = merge_models(
             str(self.params["chat_id"]),
             SpecialChatParams,
-            *source_storage_models.values(),
-            *assistant_models.values(),
+            *component_models.values(),
             config=pydantic.ConfigDict(extra="forbid"),
         )
 
         chat_params = ChatModel.model_validate(params, strict=True).model_dump(
             exclude_none=True
         )
         return {
-            component_and_action: model(**chat_params).model_dump()
-            for component_and_action, model in itertools.chain(
-                source_storage_models.items(), assistant_models.items()
-            )
+            fn: model(**chat_params).model_dump()
+            for fn, model in component_models.items()
         }
 
-    async def _enqueue(
-        self, component: Type[Component], action: str, *args: Any
-    ) -> Any:
-        try:
-            return await self._rag._queue.enqueue(
-                component, action, args, self._unpacked_params[(component, action)]
+    def _run_async(self, fn: Awaitable[T]) -> T:
```

Comment:
Ok, this whole function is a problem. We can't merge like this as it is causing issues on Windows and macOS.